### PR TITLE
make imapwildcard-branch work and with idle

### PR DESCRIPTION
**Pull request by L0ric0**

### CODE REVIEW

**Comment by girst** [comment]
thanks for fixing this up!
what's the rationale behind removing the directory if SELECT failed?

**Comment by L0ric0** [comment]
if the SELECT fails the mailbox is either not accessible or doesn't exist or the argument is invalid. At this point we don't need to check on it on further requests.

EDIT: forget it that was from only doing the `list()` request once on startup. if that should not be implemented than the else can be deleted.

I would like to do the hit on `list()` only on startup as for a account with 1000+ mailboxes we would not do this on every mail_count check.

**Comment by girst** [comment]
On Tue, Mar 31, 2020 at 06:00:37AM -0700, L0ric0 wrote:
>if the SELECT fails the mailbox is either not accessible or doesn't 
>exist or the argument is invalid. At this point we don't need to check 
>on it on further requests

but that doesn't happen regularly, right? that means, that an error has 
occured, and i would rather fail loudly. 

how 'bout this:

ok, _ = self.connection.select(directory)
if ok != "OK":
	raise imaplib.IMAP4.error("cannot select {}".format(directory))

and if you're at it, line 342 should also say `ok != "OK"` instead of 
`not ok`

finally, regarding the 3rd commit: please give a default value for the 
directories parameter like so:
	def _idle(self, directories="INBOX"):

**Comment by L0ric0** [comment]
ok i found a mailbox that has a problem with being selected because it has a whitespace will see if i can fix this

**Comment by L0ric0** [comment]
ok i noticed that the fail to select happens for me quite regularly as the parent-folders that i create with my sieve filter are not considered true mailboxes and so a select on them will throw an error. 
meaning if the sieve filter creates a mailbox named `Github/fork-graveyard/py3status` and the mailbox `GitHub/fork-graveyard` didn't exist the folder is returned by `list()` but is not selectable.

This can probably also happen if you create a mailbox with other programs or define a namespace which has no root folder.

File: py3status/modules/imap.py [hunk 1]
```diff
@@ -339,27 +338,30 @@ def _get_mail_count(self):
...
 
                 for directory in directories:
-                    self.connection.select(directory)
-                    unseen_response = self.connection.search(None, self.criterion)
-                    mails = unseen_response[1][0].split()
-                    tmp_mail_count += len(mails)
+                    ok, _ = self.connection.select('"{}"'.format(directory))
+                    if ok != "OK":
+                        if self.debug:
+                            self.py3.log("cannot select {}".format(directory))
```
Comment (L0ric0) [inline py3status/modules/imap.py:353]: i don't know if we could scream louder at this point, like marking the module with a degraded color

**Comment by girst** [comment]
just raise imaplib.IMAP4.error(), like we do in other places. this will 
do the right thing automatically then.

**Comment by L0ric0** [comment]
> just raise imaplib.IMAP4.error(), like we do in other places. this will do the right thing automatically then.
> […](#)
> On Wed, Apr 01, 2020 at 02:31:07AM -0700, L0ric0 wrote: i don't know if we could scream louder at this point, like marking the module with a degraded color

and stop the module from working as `list()` can return mailboxes that are no mailboxes. see 
https://github.com/fork-graveyard/py3status/pull/1#issuecomment-607137190

**Comment by girst** [comment]
On Wed, Apr 01, 2020 at 09:29:20AM -0700, L0ric0 wrote:
>> just raise imaplib.IMAP4.error(), like we do in other places. this will do the right thing automatically then.
>and stop the module from working as `list()` can return mailboxes that 
>are no mailboxes. see
>https://github.com/fork-graveyard/py3status/pull/1#issuecomment-607137190

sorry, missed that! we need to fix our list() parsing then! ;-)

the LIST argument has attributes that we can use to filter out the 
non-selectable mailboxes. -> 
https://tools.ietf.org/html/rfc3501#section-7.2.2

can you give me an output of list on your system, please? it's ok to 
anonymize the mailbox names. I'll cook something up!

**Comment by girst** [comment]
oops, sorry! i accidentally merged this PR. i still would like you to 
send the list() output, and for you to verify the fix.

we aren't yet done, but we're getting very close!

**Comment by girst** [comment]
@L0ric0: i've rebased the branch -- you probably want to `git pull 
--rebase` or clone afresh.

if it's ok with you, i have credited you as the author of the commit, 
with me as committer. i would like to submit this PR to upstream, but 
need you to test it out first, and report back (this merged PR-ticket is 
fine)

**Comment by L0ric0** [comment]
here is a part of the `list()` statement all others lines don't cause errors and are so much that i don't want the time to redact stuff.

The fix seems to work. I have added the `raise` again and will do another merge request.
´´´
b'(\\HasChildren \\UnMarked) "/" GitHub'                                                                                                                                                                                                                                                                                       
b'(\\Noselect \\HasChildren) "/" GitHub/fork-graveyard'                                                                                                                                                                                                                                                                        
b'(\\HasNoChildren \\UnMarked) "/" GitHub/fork-graveyard/py3status'                                                                                                                                                                                                                                                            
b'(\\Noselect \\HasChildren) "/" GitHub/nextcloud'                                                                                                                                                                                                                                                                             
b'(\\HasNoChildren \\UnMarked) "/" GitHub/nextcloud/server'                                                                                                                                                                                                                                                                    
b'(\\HasNoChildren \\UnMarked) "/" GitHub/nextcloud/contacts'                                                                                                                                                                                                                                                                  
b'(\\Noselect \\HasChildren) "/" GitHub/omarroth'                                                                                                                                                                                                                                                                              
b'(\\HasNoChildren \\UnMarked) "/" GitHub/omarroth/invidious'                                                                                                                                                                                                                                                                  
b'(\\Noselect \\HasChildren) "/" GitHub/encode'                                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/encode/httpx'                                                                                                                                                                                                                                                                        
b'(\\HasChildren \\UnMarked) "/" GitHub/turt2live'                                                                                                                                                                                                                                                                             
b'(\\HasNoChildren \\UnMarked) "/" GitHub/turt2live/matrix-media-repo'                                                                                                                                                                                                                                                         
b'(\\HasChildren \\UnMarked) "/" GitHub/mlj'                                                                                                                                                                                                                                                                                   
b'(\\HasNoChildren \\UnMarked) "/" GitHub/mlj/castget'                                                                                                                                                                                                                                                                         
b'(\\HasChildren \\UnMarked) "/" GitHub/andig'                                                                                                                                                                                                                                                                                 
b'(\\HasNoChildren \\UnMarked) "/" GitHub/andig/carddav2fb'                                                                                                                                                                                                                                                                    
b'(\\HasChildren \\UnMarked) "/" GitHub/transmission'                                                                                                                                                                                                                                                                          
b'(\\HasNoChildren \\UnMarked) "/" GitHub/transmission/transmission'                                                                                                                                                                                                                                                           
b'(\\HasChildren \\UnMarked) "/" GitHub/raithos'                                                                                                                                                                                                                                                                               
b'(\\HasNoChildren \\UnMarked) "/" GitHub/raithos/xwing'                                                                                                                                                                                                                                                                       
b'(\\HasChildren \\UnMarked) "/" GitHub/hevelius'                                                                                                                                                                                                                                                                              
b'(\\HasNoChildren \\UnMarked) "/" GitHub/hevelius/files_external_onedrive'                                                                                                                                                                                                                                                    
b'(\\HasChildren \\UnMarked) "/" GitHub/NARKOZ'                                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/NARKOZ/hacker-scripts'                                                                                                                                                                                                                                                               
b'(\\HasChildren \\UnMarked) "/" GitHub/poljar'                                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/poljar/weechat-matrix'                                                                                                                                                                                                                                                               
b'(\\HasChildren \\UnMarked) "/" GitHub/matrix-org'                                                                                                                                                                                                                                                                            
b'(\\HasNoChildren) "/" GitHub/matrix-org/synapse'                                                                                                                                                                                                                                                                             
b'(\\HasChildren \\UnMarked) "/" GitHub/Kwpolska'                                                                                                                                                                                                                                                                              
b'(\\HasNoChildren \\UnMarked) "/" GitHub/Kwpolska/pkgbuilder'                                                                                                                                                                                                                                                                 
b'(\\HasChildren \\UnMarked) "/" GitHub/actionless'                                                                                                                                                                                                                                                                            
b'(\\HasNoChildren \\UnMarked) "/" GitHub/actionless/pikaur'                                                                                                                                                                                                                                                                   
b'(\\HasChildren \\UnMarked) "/" GitHub/tulir'                                                                                                                                                                                                                                                                                 
b'(\\HasNoChildren \\UnMarked) "/" GitHub/tulir/mautrix-python'                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/tulir/mautrix-telegram'                                                                                                                                                                                                                                                              
b'(\\HasChildren \\UnMarked) "/" GitHub/cylgom'                                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/cylgom/ly'                                                                                                                                                                                                                                                                           
b'(\\HasChildren \\UnMarked) "/" GitHub/ultrabug'                                                                                                                                                                                                                                                                              
b'(\\HasNoChildren) "/" GitHub/ultrabug/py3status'                                                                                                                                                                                                                                                                             
b'(\\HasChildren \\UnMarked) "/" GitHub/pstauffer'                                                                                                                                                                                                                                                                             
b'(\\HasNoChildren \\UnMarked) "/" GitHub/pstauffer/gitlab-webhook-receiver'                                                                                                                                                                                                                                                   
b'(\\HasChildren \\UnMarked) "/" GitHub/maubot'                                                                                                                                                                                                                                                                                
b'(\\HasNoChildren \\UnMarked) "/" GitHub/maubot/reminder'                                                                                                                                                                                                                                                                     
b'(\\HasNoChildren \\UnMarked) "/" GitHub/maubot/rss'                                                                                                                                                                                                                                                                          
b'(\\HasNoChildren \\UnMarked) "/" GitHub/maubot/maubot'                                                                                                                                                                                                                                                                       
b'(\\HasNoChildren \\UnMarked) "/" GitHub/maubot/gitlab'                                                                                                                                                                                                                                                                       
b'(\\HasChildren \\UnMarked) "/" GitHub/python-gitlab'                                                                                                                                                                                                                                                                         
b'(\\HasNoChildren \\UnMarked) "/" GitHub/python-gitlab/python-gitlab'                                                                                                                                                                                                                                                         
b'(\\HasChildren \\UnMarked) "/" GitHub/xournalpp'                             
b'(\\HasNoChildren \\UnMarked) "/" GitHub/xournalpp/xournalpp'
b'(\\HasChildren \\UnMarked) "/" GitHub/tremc'                                 
b'(\\HasNoChildren \\UnMarked) "/" GitHub/tremc/tremc'
```

**Comment by girst** [comment]
On Thu, Apr 02, 2020 at 01:16:24AM -0700, L0ric0 wrote:
>and are so much that i don't want the time to redact stuff.

no worries, the list you've provided is more than enough (and I believe 
our fix is correct)

>The fix seems to work.

great to hear!

>I have added the `raise` again and will do another merge request.

I don't see any changes compared to fork-graveyard.imapwildcard[1], so
i've opened the PR as https://github.com/ultrabug/py3status/pull/1926.  
since your name shows up as the commit author, you'll be the one 
credited once this lands.

Thanks for this feature! i feel we landed on an implementation that 
really benefits py3status (and hopefully, you're happy with it too).

[1]: https://github.com/fork-graveyard/py3status/compare/imapwildcard..L0ric0:imapwildcard

**Comment by L0ric0** [comment]
> I have added the `raise` again and will do another merge request.
> I don't see any changes compared to fork-graveyard.imapwildcard[1], so i've opened the PR as [ultrabug#1926](https://github.com/ultrabug/py3status/pull/1926). since your name shows up as the commit author, you'll be the one credited once this lands. Thanks for this feature! i feel we landed on an implementation that really benefits py3status (and hopefully, you're happy with it too). [1]: https://github.com/fork-graveyard/py3status/compare/imapwildcard..L0ric0:imapwildcard

after the rebase on my branch the loglines were still present. After i saw that you had changed that already I had forgotten about the comment.

**Pull request merged**